sympy/codegen/ast.py

```diff
@@ -0,0 +1,627 @@
+
+class Assignment(Relational):
+    """
+    Represents variable assignment for code generation.
+
+    Parameters
+    ----------
+    lhs : Expr
+        Sympy object representing the lhs of the expression. These should be
+        singular objects, such as one would use in writing code. Notable types
+        include Symbol, MatrixSymbol, MatrixElement, and Indexed. Types that
+        subclass these types are also supported.
+
+    rhs : Expr
+        Sympy object representing the rhs of the expression. This can be any
+        type, provided its shape corresponds to that of the lhs. For example,
+        a Matrix type can be assigned to MatrixSymbol, but not to Symbol, as
+        the dimensions will not align.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, MatrixSymbol, Matrix
+    >>> from sympy.codegen.ast import Assignment
+    >>> x, y, z = symbols('x, y, z')
+    >>> Assignment(x, y)
+    Assignment(x, y)
+    >>> Assignment(x, 0)
+    Assignment(x, 0)
+    >>> A = MatrixSymbol('A', 1, 3)
+    >>> mat = Matrix([x, y, z]).T
+    >>> Assignment(A, mat)
+    Assignment(A, Matrix([[x, y, z]]))
+    >>> Assignment(A[0, 1], x)
+    Assignment(A[0, 1], x)
+    """
+
+    rel_op = ':='
+    __slots__ = []
+
+    def __new__(cls, lhs, rhs=0, **assumptions):
+        from sympy.matrices.expressions.matexpr import (
+            MatrixElement, MatrixSymbol)
+        from sympy.tensor.indexed import Indexed
+        lhs = _sympify(lhs)
+        rhs = _sympify(rhs)
+        # Tuple of things that can be on the lhs of an assignment
+        assignable = (Symbol, MatrixSymbol, MatrixElement, Indexed)
+        if not isinstance(lhs, assignable):
+            raise TypeError("Cannot assign to lhs of type %s." % type(lhs))
+        # Indexed types implement shape, but don't define it until later. This
+        # causes issues in assignment validation. For now, matrices are defined
+        # as anything with a shape that is not an Indexed
+        lhs_is_mat = hasattr(lhs, 'shape') and not isinstance(lhs, Indexed)
+        rhs_is_mat = hasattr(rhs, 'shape') and not isinstance(rhs, Indexed)
+        # If lhs and rhs have same structure, then this assignment is ok
+        if lhs_is_mat:
+            if not rhs_is_mat:
+                raise ValueError("Cannot assign a scalar to a matrix.")
+            elif lhs.shape != rhs.shape:
+                raise ValueError("Dimensions of lhs and rhs don't align.")
+        elif rhs_is_mat and not lhs_is_mat:
+            raise ValueError("Cannot assign a matrix to a scalar.")
+        return Relational.__new__(cls, lhs, rhs, **assumptions)
+
+# XXX: This should be handled better
+Relational.ValidRelationOperator[':='] = Assignment
+
+# The following are defined to be sympy approved nodes. If there is something
+# smaller that could be used, that would be preferable. We only use them as
+# tokens.
+
+
+class NativeOp(with_metaclass(Singleton, Basic)):
```

Comment:
Is Singleton really needed here? Just because an object is atomic (no args) doesn't mean it needs to be singletonized. `op == AddOp()` or `isinstance(op, AddOp)` should work regardless, because of the way equality is defined on Basic. The only thing Singleton enables is `is` comparison, which is a bit of an antipattern unless it's needed for performance or memory purposes (which it really isn't here).